More stuff

ezyang opened the pull request:
More stuff

Summary: boom

Test Plan: boom

Differential Revision: https://phabricator.intern.facebook.com/D9514690

zdevito commented on torch/csrc/autograd/variable.h:
```diff
@@ -8,7 +8,6 @@
 
 #include "torch/csrc/autograd/function.h"
 #include "torch/csrc/autograd/variable_version.h"
-#include "torch/csrc/jit/ir.h"
```
Removing this saves a ton of compilation time because of how many files include variable.h

zdevito commented:
This can now produce fusion blocks like this for the LSTM:

```
with fusion_group_6 = graph(%13, %15, %17, %19, %21) {
  %22 = Sigmoid!(%21), uses = [%11.i0];
  %20 = Sigmoid!(%19), uses = [%14.i0];
  %18 = Tanh!(%17), uses = [%11.i1];
  %16 = Sigmoid!(%15), uses = [%3.i0];
  %14 = Mul!(%20, %13), uses = [%8.i0];
  %11 = Mul!(%22, %18), uses = [%8.i1];
  %8 = Add!(%14, %11), uses = [%5.i0, %0.i1];
  %5 = Tanh!(%8), uses = [%3.i1];
  %3 = Mul!(%16, %5), uses = [%0.i0];
  return (%3, %8);
}
```

Still needs cleanup and comments. Mutable IR is a double-edges sword. It is really helpful because you can always see intermediate states of a pass, but tasks like incrementally creating FusionGroups is pretty verbose.

ezyang commented on torch/csrc/jit/graph_fuser.cpp:
```diff
@@ -0,0 +1,191 @@
+  "Negate",
+};
+
+struct GraphFuser {
```
Let's either insert explicit lint calls once all invariants are restored after a "pass", or figure out how to instrument this automatically (we have some optimizer)

zdevito replied:
Explicitly calling verify is good. I'd like to avoid a PassManager, that is a disaster in LLVM.

ezyang commented on torch/csrc/jit/ir.h:
```diff
@@ -69,29 +64,48 @@ _(PythonOp) \
+  friend struct graph_node_list_iterator;
+  friend struct graph_node_list;
 private:
+  // each node but Return/Param
```
This seems inconsistent with the later statement that the Return node is used as the sentinel for the beginning and end of the list. In that case, isn't the Return node also in the list?

zdevito replied:
Return  acts as graph_node_iterator.end(), so it is, in fact, not in the list. Given that we are going to remove this implementation of lists in favor of std::list, this inconsistency will go away.

ezyang commented on torch/csrc/jit/ir.h:
```diff
@@ -69,29 +74,48 @@ _(PythonOp) \
+  // next_in_graph[0] is next pointer
+  // next_in_graph[1] is prev pointer
+  // using an array to allow the same iterator class for forward and reverse node lists
+  Node * next_in_graph[2];
```
Is this list in topological order? Based on how `original_topological_position` is computed the answer seems to be yes.

ezyang replied:
From chat, the answer is yes.

ezyang commented:
So, suppose I have:

```
%1 = f(%2);
%3 = g(%4)
```

where the order I list these instructions is the topological sort. Is it legal to call `%1.addInput(%3)`? My guess is no. In that case, how are we actually supposed to do the graph motion necessary to make this call legal?

ezyang commented:
@apaszke Re std::list versus hand-rolled list pointer, http://llvm.org/docs/ProgrammersManual.html#list in the LLVM programmer manuals has some justification for why they didn't use `std::list`. (I haven't understood the argument enough to know if I believe it.)

ezyang commented on torch/csrc/jit/ir.h:
```diff
@@ -69,29 +74,48 @@ _(PythonOp) \
+  // is associated with exactly one place in the node list...
+  // of the graph_
+  // this circular is a doubly-linked list, the Return node is used as the sentinel for the beginning and end of the list
+  // such that the list never has null pointers
```
If I understand correctly, this sentinel is used so that we have a valid `end()` iterator?

zdevito replied:
1. It's a valid end iterator, which is nice.
2. (more important) There are never any null pointers in prev/next pointers when something is in the list. So it is _always_ valid to traverse prev/next. 

I could have used another specific sentinel type for the list but then it would appear as a separate Node type which is annoying in its own way. Using std::list should simplify things.

zdevito commented:
> So, suppose I have:
> 
> %1 = f(%2);
> %3 = g(%4)
> where the order I list these instructions is the topological sort. Is it legal to call %1.addInput(%3)? My guess is no. Does this situation ever come up?
> 
You can first use %3.moveBefore(%1) first and then add the node, or move %1 after %3. Typically when making change you know which of those two would maintain topological order.

ezyang commented:
Observation: linter is going to need to check that (1) forward/backward list links are consistent (non-issue if we use `std::list`), (2) topological ordering is respected by inputs/outputs. Maybe other stuff which I haven't noticed yet.

apaszke commented on torch/csrc/jit/graph_fuser.cpp:
```diff
@@ -0,0 +1,191 @@
+          // replaces the Select(PythonOp,0) with just SimpleMap(opname)
+          auto new_op = graph->create<SimpleMap>(name,p->inputs());
+          new_op->insertAfter(p);
+          JIT_ASSERT(1 == p->uses().size());
```
Maybe add `&& p->cast<Select>()`?

apaszke commented on torch/csrc/jit/ir.h:
```diff
@@ -114,180 +129,495 @@ struct Node {
+  //          %5 = h(%1)
+  //          %4 = g(%3)
+  void insertBefore(Node * n) {
+    JIT_ASSERT(n->inGraphList());
```
I'd add `!inGraphList()`

apaszke commented on torch/csrc/jit/ir.h:
```diff
@@ -114,180 +129,495 @@ struct Node {
+  //        %3 = g(%1)
+  // Execute: %2.eraseFromParent()
+  // Result: %3 = g(%1)
+  void eraseFromParent();
```
Can we please change the name to `eraseFromGraph` or even `dealloc` to indicate that it should never be used again? There's no such thing as parent node as far as I know, so this name is very confusing

zdevito replied:
eraseFromParent comes from llvm. I'd rather not change it, mostly because it aligns with LLVM so it has a familiar naming scheme.  While there aren't parents _now_, if we every add control flow and basic blocks, we are going to be erasing from Nodes from BasicBlocks and BasicBlocks from Graphs.

apaszke commented on torch/csrc/jit/ir.h:
```diff
@@ -114,180 +129,495 @@ struct Node {
+    return old_node;
+  }
+  bool inGraphList() {
+    return next() && prev();
```
maybe add `JIT_ASSERT(!(next() ^ prev())); return next();`

apaszke commented on torch/csrc/jit/ir.h:
```diff
@@ -114,180 +129,495 @@ struct Node {
+    Node * next = this->next();
+    Node * prev = this->prev();
+    prev->next() = next;
+    next->prev() = prev;
```
nit: there's no need to assign them to locals.

apaszke commented on torch/csrc/jit/ir.h:
```diff
@@ -114,180 +129,495 @@ struct Node {
+template<typename Self, NodeKind K>
+struct Primitive : public NodeWithKind<Self,K> {
+  void init() {}
+  void init(ArrayRef<Node*> inputs) {
```
What's an ArrayRef? Do we really want to add dependency on ATen here?

zdevito replied:
Taken from LLVM, It is a view of contiguous memory vectors so you can write stuff like:

init({3,4})

and not have it allocate a vector. Other solutions require overloading init for std::initializer_list, which gets more verbose. We can simply copy ArrayRef.h into jit/ to avoid the dependency if you wish.

apaszke commented on torch/csrc/jit/ir.h:
```diff
@@ -114,180 +129,495 @@ struct Node {
 
   // only used to keep track of allocated nodes
   // actual representation of Graph is done with
   // inputs, outputs, nodes
```
`nodes_` no longer exists

apaszke commented on torch/csrc/jit/ir.h:
```diff
@@ -114,180 +129,495 @@ struct Node {
-    return r;
-  }
+  //allows fast removal of nodes when they are deleted.
+  std::unordered_set<Node*> all_nodes;
```
Is there any reason why we don't want to have Nodes as values of this set? Do we ever move them between graphs? I don't think so

zdevito replied:
We don't do that now, but this is something that LLVM does allow. Eventually nodes will live in basic blocks, and it will be common for a node to move between basic blocks inside a function, so putting them inside the basic block's std::list would not be a good idea. I don't want to preclude these changes by the design of the ownership. Finally, there are a few nodes in the Graph like Param/Return nodes that do not exist in this list, and we might want to, for instance, reorder Param nodes while a map from Node* -> value is still live.

apaszke commented on torch/csrc/jit/ir.h:
```diff
@@ -335,4 +665,97 @@ static inline const char * toString(NodeKind kind) {
+    this->is_legacy = is_legacy;
+  }
+  virtual void cloneFrom(PythonOp * other) override {
+    throw std::runtime_error("cannot clone PythonOp because of THPObjectPtr");
```
Why is that a problem? Just `Py_INCREF` the objects and give the pointer to another `THPObjectPtr`

apaszke commented on torch/csrc/jit/graph_fuser.cpp:
```diff
@@ -0,0 +1,191 @@
+  std::unique_ptr<Graph> graph;
+  // used to order nodes so we alway consider producer-consumer fusions
+  // in reverse topological order
+  std::unordered_map<Node*,size_t> original_topological_position;
```
Maybe just `topo_idx`? Not sure why is it an original position, as some nodes are inserted while the fuser works

apaszke commented on torch/csrc/jit/graph_fuser.cpp:
```diff
@@ -0,0 +1,191 @@
+    size_t i = 0;
+    for(auto input : group->inputs()) {
+      inputs_map[input] = subgraph.inputs()[i++];
+    }
```
Is it an invariant that `group->inputs().size() == subgraph.inputs().size()`?

zdevito replied:
yes, ill add an assert here

apaszke commented:
It would also be great if instead of adding more and more debugging functions, we started writing unit tests. This makes it much easier to work with others' code while being sure that you at least don't break simple things

zdevito commented on torch/csrc/jit/graph_fuser.cpp:
```diff
@@ -0,0 +1,191 @@
+
+  GraphFuser(std::unique_ptr<Graph> graph)
+  : graph(std::move(graph)) {}
+  void replacePythonOps() {
```
Yeah, you can put this in the tracing logic. I added a note to the current code saying that it should be moved.

zdevito commented:
I updated the PR with the fixes that were mentioned. Adam, if you are ok with the comments, go ahead and merge.

apaszke commented:
Looks good now :+1: I'll rebase my PR tomorrow

zdevito commented:
I agree about unit tests. I'd rather not expose the entire IR to python (it's going to be big, it's going to change, lots of stuff to refactor if we expose it), but we can use the trace recorder to get IR from normal pytorch code, and then expose functions that work at the graph level. We can discuss tomorrow. Any change that we can just write unit tests directly in C++? It would certainly be more straightforward for some of our components.

apaszke commented:
I'd prefer to keep them in Python, as this is what we use for contbuilds at the moment, but we can add a temporary .cpp file that we can use during development, and translate it into Python once we're pretty sure the IR will mostly be stable

zdevito commented:
Ok, I think having a temporary cpp file is good for now. Is it bad to expose a single python function jit._run_all_tests() that just runs our test suite from Python? I imagine the binary size increase from including this code would be minor.

apaszke commented:
Ah I see. No, this seems ok to me

ezyang commented:
@apaszke Thanks for reminding us on tests :) I'm pretty keen on us having an accept-style test system for the E2E tests, so I will bang on that some today.

ezyang committed:
Trivial change arf

Summary: none

Differential Revision: https://phabricator.intern.facebook.com/D9514646

ezyang committed:
More stuff

Summary: boom

Test Plan: boom

Differential Revision: https://phabricator.intern.facebook.com/D9514690
